cmd/rdma/main.go

```diff
@@ -1,10 +1,261 @@
+	if err != nil {
+		return fmt.Errorf("failed to open network namespace %s: %v", nsPath, err)
+	}
+	defer sourceNs.Close()
+
+	targetNs, err := plugin.nsm.GetCurrentNS()
+	if err != nil {
+		return fmt.Errorf("failed to open current network namespace: %v", err)
+	}
+	defer targetNs.Close()
+
+	err = sourceNs.Do(func(_ ns.NetNS) error {
+		// Move RDMA device to default namespace
+		log.Printf("INFO: cmdDel: Moving rdmaDev %s from container namespace to current namespace", rdmaDev)
+		return plugin.rdmaManager.MoveRdmaDevToNs(rdmaDev, targetNs)
+	})
+	if err != nil {
+		return fmt.Errorf("failed to move RDMA device %s to default namespace. %v", rdmaDev, err)
+	}
+	return err
+}
+
+func (plugin *rdmaCniPlugin) CmdAdd(args *skel.CmdArgs) error {
+	log.Printf("INFO: cmdAdd: args: %+v ", args)
+	conf, err := plugin.parseConf(args.StdinData, args.Args)
+	if err != nil {
+		return err
+	}
+
+	// Ensure RDMA-CNI was called as part of a chain, and parse PrevResult
+	if conf.RawPrevResult == nil {
+		return fmt.Errorf("RDMA-CNI is expected to be called as part of a plugin chain")
+	}
+	if err := version.ParsePrevResult(&conf.NetConf); err != nil {
+		return err
+	}
+	result, err := current.NewResultFromResult(conf.PrevResult)
+	if err != nil {
+		return err
+	}
+	log.Printf("DEBUG: prev results: %+v", result)
+
+	// Ensure RDMA subsystem mode
+	err = plugin.ensureRdmaSystemMode()
+	if err != nil {
+		return err
+	}
+
+	// Delegate plugin may not add Device ID to the network configuration, if so,
+	// attempt to derive it from PrevResult Mac address with some sysfs voodoo
+	if conf.DeviceID == "" {
+		if conf.DeviceID, err = plugin.deriveDeviceIdFromResult(result); err != nil {
+			return err
+		}
+	}
+
+	rdmaDevs, err := plugin.rdmaManager.GetRdmaDevsForPciDev(conf.DeviceID)
+	if err != nil || len(rdmaDevs) == 0 {
+		// TODO: Should we just return nil here ?
+		return fmt.Errorf("failed to get RDMA devices for PCI device: %s. %v", conf.DeviceID, err)
+	}
+
+	if len(rdmaDevs) != 1 {
+		// Expecting exactly one RDMA device
+		return fmt.Errorf(
+			"discovered more than one RDMA device %v for PCI device %s. Unsupported state", rdmaDevs, conf.DeviceID)
+	}
+
+	// Move RDMA devices to container namespace
+	rdmaDev := rdmaDevs[0]
+	log.Printf("INFO: moving RDMA device %s to namespace %s", rdmaDev, args.Netns)
+
+	err = plugin.moveRdmaDevToNs(rdmaDev, args.Netns)
+	if err != nil {
+		return fmt.Errorf("failed to move RDMA device %s to namespace. %v", rdmaDev, err)
+	}
+
+	// Save RDMA state
+	state := rdmatypes.NewRdmaNetState()
+	state.DeviceID = conf.DeviceID
+	state.SandboxRdmaDevName = rdmaDev
+	state.ContainerRdmaDevName = rdmaDev
+	pRef := plugin.stateCache.GetStateRef(conf.Name, args.ContainerID, args.IfName)
+	err = plugin.stateCache.Save(pRef, &state)
+	if err != nil {
+		// Move RDMA dev back to current namespace
+		restoreErr := plugin.moveRdmaDevFromNs(state.ContainerRdmaDevName, args.Netns)
+		if restoreErr != nil {
```

Comment:
you won't print why it error out before the restore error. Maybe just add log print in line 204